async/gluten_async_intf.ml

```diff
@@ -82,4 +82,5 @@ module type Client = sig
   val upgrade : _ t -> Gluten.impl -> unit
   val shutdown : _ t -> unit Deferred.t
   val is_closed : _ t -> bool
+  val close_finished : _ t -> unit Deferred.t
```

Comment:
Instead of introducing a new API, could we make this part of the `val shutdown: _ t -> unit Deferred.t` process?

Reply:
To be used as a functor for `Persistent_connection_kernel.Make`, it should follow this interface:

```ocaml
module type Closable = sig
  (** a connection type *)
  type t

  (** [close t] closes the connection. The returned deferred becomes determined once any
      resources needed to maintain the connection have been released. *)
  val close : t -> unit Deferred.t

  (** [is_closed t] returns true if [close] has ever been called (even if the returned
      deferred has not yet been fulfilled).

      Note that some modules implementing [Closable] may call close internally upon
      noticing that the connection was closed by the other side. The interface of such a
      module ought to say that this is the case. *)
  val is_closed : t -> bool

  (** [close_finished t] becomes determined at the same time as the result of the first
      call to [close]. [close_finished] differs from [close] in that it does not have the
      side effect of initiating a close. *)
  val close_finished : t -> unit Deferred.t
end
```

Reply:
Actually, since there is a parameter it cannot be used directly as an argument to `Persistent_connection_kernel.Make`. I used this patch in an internal fork to `ocaml-h2` that I could maybe upstream at some point.

Reply:
But as the doc says, `close` and `close_finished` have not the same semantics.
Also, would not be better to call `shutdown` -> `close` since the socket is closed in the process?

Reply:
- leaving this here for my future reference: https://github.com/janestreet/async_kernel/blob/9f747452c88fe716cda4bd6f29ef0f3cb1ff3a39/persistent_connection_kernel/src/persistent_connection_kernel_intf.ml#L15


> would not be better to call shutdown -> close since the socket is closed in the process?

works for me, would you like to propose a separate PR for that?